### DISCUSSION TITLE
v7 with Next 13 and Turborepo - MantineProvider Issue

### DISCUSSION THREAD

**Discussion by benjamin-musil** [mantinedev/mantine#4718] [Q&A] [answered]
Edit: I downgraded back to v6 to double check and I must have done something different while I was migrating, since I get the same issue in v6 as well. So this is really a question about how to share the same MantineProvider across a monorepo.

Relevant parts of original post:
I'm having an issue with importing components from my ui package (which consumes mantine) into my nextjs app within the monorepo. My nextjs app _also_ consumes mantine. Part of the structure is as follows:

```
turborepo-root
├── apps
│   ├── my-app
|       ├──app
|           ├──layout.tsx
├── packages
│   ├── ui
│           └── MyComponent.tsx
```

In this case, `MyComponent` wrapped a mantine component and then exported to `my-app`. Under the `layout` component, it used the `MantineProvider` installed in the nextjs app. Like so:

```jsx
import { MantineProvider, ColorSchemeScript } from "@mantine/core";
import { MyComponent } from "ui";

export default function RootLayout({
  children,
}: {
  children: React.ReactNode;
}) {
  return (
    <html lang="en">
      <head>
        <link rel="icon" href="/favicon.ico" sizes="any" />
        <ColorSchemeScript />
      </head>
      <body>
        <MantineProvider>
            <MyComponent />
        </MantineProvider>
      </body>
    </html>
  );
}
```

This was producing the following error:
```
Unhandled Runtime Error
Error: @mantine/core: MantineProvider was not found in component tree, make sure you have it in your app
```

If I don't use a component from my `ui` package, I can see in the dev tools that the `MantineProvider` is definitely in the component tree. I tried moving my `ui` package's mantine dependency to the peer dependency group to no effect. I was able to finally get it to work by exporting a `MantineProvider` from the `ui` package and not using any mantine directly imported in `my-app`. Not sure if I'm doing something wrong or if this is actually a bug. I'd like to be able to install and use mantine both in my `ui` package and my app (since I will have multiple apps as well, being this is a monorepo).

**Comment by rtivital** [comment, marked as answer]
Seems like each of your packages have their own copy of `@mantine/core` bundled. To fix that, make sure that you only have one copy of `@mantine/core` for all packages.

**Reply by benjamin-musil** [reply]
Putting my resolution here for future internet travelers because I spent a *lot* of time on Google if anyone else ran into this. The reason is indeed multiple instances of mantine, and this was how I fixed for myself:

If you're using turborepo or another monorepo tool, make sure you also do a clean wipe of the root `node_modules` after you mark the shared package with a peer dependency for mantine, and the consuming app(s) as a normal dependency.

**Reply by AndrewJSchoen** [reply]
@benjamin-musil Did you have to put the peer dependencies in each of your consumed component libraries, or at the top-level of the turbo repo? In other words, right now I have the mantine packages set as peer dependencies in my root `package.json`, and then in my storybook and app packages, they are set as dependencies. Unfortunately, the error still persists when configured that way.

**Reply by jonorogers** [reply]
@rtivital @benjamin-musil Could you guys share a bit more about the specifics of this fix? 

I have two projects; frontend and ui-kit. I want to define components using mantine in ui-kit and import them into frontend, and I want to use mantine components directly in frontend. Seems like it should be fairly simple, no?

mantine is a dependency of frontend. 

If it's set as a peerDependency of ui-kit, the app won't compile - vite complains that the dependency is imported but could not be resolved. 
If it's set as a dependency of ui-kit, the app will compile - but then at runtime I get the fun 'MantineProvider was not found in component tree'.

I'm using yarn workspaces (v4.9.2), typescript project references, and vite aliases.

**Reply by jonorogers** [reply]
For anyone else who stumbles upon this issue; I solved it by making the mantine packages (in my ui-kit shared library) both peerDependenices and devDependencies. I'm not 100% sure this is the right approach, but it's working for me.
